Mathlib/Topology/Algebra/InfiniteSum/ENNReal.lean

```diff
@@ -0,0 +1,61 @@
+/-
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Data.Real.ENatENNReal
+import Mathlib.Data.Set.Card
+import Mathlib.Topology.Instances.ENNReal
+
+/-!
+# Infinite sums of ENNReal and Set.encard
+
+This file provides lemmas relating sums of constants to the cardinality of the domain of these sums.
+
+-/
+
+open Set Function
+
+namespace ENNReal
+
+lemma tsum_one_eq' {α : Type*} (s : Set α) : ∑' (_:s), (1 : ℝ≥0∞) = s.encard := by
+  if hfin : s.Finite then
+    have hfin' : Finite s := by exact hfin
+    rw [tsum_def]
+    simp only [ENNReal.summable, ↓reduceDIte]
+    have hsup: support (fun (_ : s) ↦ (1 : ℝ≥0∞)) = Set.univ := by
+      ext i
+      simp only [mem_support, ne_eq, one_ne_zero, not_false_eq_true, mem_univ]
+    have hsupfin: (Set.univ : Set s).Finite := by exact finite_univ
+    rw [← hsup] at hsupfin
+    rw [if_pos hsupfin]
+    rw [hfin.encard_eq_coe_toFinset_card]
+    simp only [ENat.toENNReal_coe]
+    rw [Finset.card_eq_sum_ones]
+    rw [finsum_eq_sum (fun (_ : s) ↦ (1 :ℝ≥0∞)) hsupfin]
+    simp only [Finset.sum_const, nsmul_eq_mul, mul_one, smul_eq_mul, Nat.cast_inj]
+    apply Finset.card_bij (fun a _ => a.val)
+    · intro a
+      simp only [Finite.mem_toFinset, mem_support, ne_eq, one_ne_zero, not_false_eq_true,
+        Subtype.coe_prop, imp_self]
+    · intro a _ a' _ heq
+      ext
+      exact heq
+    · intro a ha
+      use ⟨a,by
+        simp only [Finite.mem_toFinset] at ha
+        exact ha⟩
+      simp only [Finite.mem_toFinset, mem_support, ne_eq, one_ne_zero, not_false_eq_true,
+        exists_const]
+  else
+  have : Infinite s := by exact infinite_coe_iff.mpr hfin
+  rw [tsum_const_eq_top_of_ne_zero (by norm_num)]
+  rw [Set.encard_eq_top_iff.mpr hfin]
+  simp only [ENat.toENNReal_top]
+
+lemma tsum_const_eq' {α : Type*} (s : Set α) (c : ℝ≥0∞) :
+    ∑' (_:s), (c : ℝ≥0∞) = s.encard * c := by
+  nth_rw 1 [← one_mul c]
+  rw [ENNReal.tsum_mul_right,tsum_one_eq']
```

Comment:
Similar ideas, but note how we're now using the `simp` lemma from above
```suggestion
@[simp]
lemma tsum_set_const_eq (c : ℝ≥0∞) :
    ∑' (_ : s), (c : ℝ≥0∞) = s.encard * c := by
  rw [← one_mul c, ENNReal.tsum_mul_right]
  simp
```

Reply:
surely not having a simp makes the proof more robust?

Reply:
In this particular case, it's not really a big deal, you can leave it how you had it if you prefer.

However, to answer your question, I would say two things:

1. Given the choice between say `rw [lem1, lem2, lem3, lem4, lem5]` and `simp`, I would greatly prefer the latter for readability and proof comprehension.
2. Given the same choice, I would again prefer `simp` (as a *closing* tactic only) for robustness, because it is not sensitive to changes in lemma names. Name changes are much more common than moving a given lemma into or out of the `simp` set.

In this case, I was opting for `simp` because it conveniently avoided the need to do an `nth_rw` (because even though it rewrites the rhs too, that gets `simp`ed away). But again, the proof you had here was fine.